File: app/providers/external_http.py
from __future__ import annotations

import httpx

from app.providers.base import TweetData, TweetProvider, TweetProviderError


class ExternalHttpTweetProvider(TweetProvider):
    def __init__(self, base_url: str, *, api_key: str | None = None, timeout: float = 10.0) -> None:
        self._api_key = api_key
        self._client = httpx.AsyncClient(base_url=base_url.rstrip("/"), timeout=timeout)

    async def get_tweet(self, tweet_id: str, source_url: str) -> TweetData:
        headers = {}
        if self._api_key:
            headers["Authorization"] = f"Bearer {self._api_key}"

        try:
            response = await self._client.get(
                f"/tweets/{tweet_id}",
                params={"url": source_url},
                headers=headers,
            )
            if response.status_code == 404:
                raise TweetProviderError("tweet not found", code="not_found")
            if response.status_code == 401:
                raise TweetProviderError("provider authentication failed", code="provider_auth")
            if response.status_code == 429:
                raise TweetProviderError(
                    "provider rate limit exceeded", code="provider_rate_limited"
                )
            response.raise_for_status()
        except TweetProviderError:
            raise
        except httpx.HTTPError as exc:
            raise TweetProviderError(str(exc), code="provider_http_error") from exc

        payload = response.json()
        return TweetData.from_payload(payload.get("tweet", payload))

    async def health(self) -> bool:
        try:
            response = await self._client.get("/health")
            return response.status_code < 500
        except httpx.HTTPError:
            return False

    async def close(self) -> None:
        await self._client.aclose()

File: src/providers/externalHttp.ts
import {
  tweetFromPayload,
  TweetProviderError,
  type TweetData,
  type TweetDataPayload,
  type TweetProvider,
} from "@/providers/base";
import { buildUrl, getFetch, withTimeout, type FetchLike } from "@/providers/http";

export class ExternalHttpTweetProvider implements TweetProvider {
  private readonly baseUrl: string;
  private readonly apiKey: string | null;
  private readonly timeoutMs: number;
  private readonly fetchImpl: FetchLike;

  constructor(
    baseUrl: string,
    options: { apiKey?: string | null; timeoutSeconds?: number; fetch?: FetchLike } = {},
  ) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.apiKey = options.apiKey ?? null;
    this.timeoutMs = Math.round((options.timeoutSeconds ?? 10) * 1000);
    this.fetchImpl = getFetch(options.fetch);
  }

  async getTweet(tweetId: string, sourceUrl: string): Promise<TweetData> {
    const headers: Record<string, string> = { Accept: "application/json" };
    if (this.apiKey) headers["Authorization"] = `Bearer ${this.apiKey}`;
    const url = buildUrl(this.baseUrl + "/", `tweets/${encodeURIComponent(tweetId)}`, {
      url: sourceUrl,
    });

    const { signal, clear } = withTimeout(this.timeoutMs);
    let response: Response;
    try {
      response = await this.fetchImpl(url, { headers, signal });
    } catch (error) {
      if ((error as Error).name === "AbortError") {
        throw new TweetProviderError("request timed out", { code: "provider_timeout" });
      }
      throw new TweetProviderError(String(error), { code: "provider_http_error" });
    } finally {
      clear();
    }

    if (response.status === 404) {
      throw new TweetProviderError("tweet not found", { code: "not_found" });
    }
    if (response.status === 401) {
      throw new TweetProviderError("provider authentication failed", { code: "provider_auth" });
    }
    if (response.status === 429) {
      throw new TweetProviderError("provider rate limit exceeded", {
        code: "provider_rate_limited",
      });
    }
    if (!response.ok) {
      throw new TweetProviderError(`provider HTTP ${response.status}`, {
        code: "provider_http_error",
      });
    }

    let payload: unknown;
    try {
      payload = await response.json();
    } catch (error) {
      throw new TweetProviderError(String(error), { code: "provider_bad_response" });
    }
    const obj = payload as { tweet?: TweetDataPayload } & TweetDataPayload;
    const data: TweetDataPayload = (obj.tweet ?? obj) as TweetDataPayload;
    return tweetFromPayload(data);
  }

  async health(): Promise<boolean> {
    try {
      const { signal, clear } = withTimeout(this.timeoutMs);
      try {
        const response = await this.fetchImpl(this.baseUrl + "/health", { signal });
        return response.status < 500;
      } finally {
        clear();
      }
    } catch {
      return false;
    }
  }

  async close(): Promise<void> {
    // no-op
  }
}
